### PR TITLE
nixos/plymouth: Add label plugin and a font to the initrd

### LINKED ISSUE #59992: plymouth shows NixOS version only at shutdown, not at start (when GRUB is used)
## Issue description

It should always show the NixOS version.

Reproducible on:

- Lenovo Thinkpad X230 with **grub**

Not reproducible on:

- Apple iMac 11,3 with systemd-boot (EFI)
- Thinkpad L470 with systemd-boot (EFI)

### Steps to reproduce

1. boot system
2. shut down system

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.36, NixOS, 19.03.172311.454eea84a75 (Koi)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(davidak): `""`
 - channels(root): `"nixos-19.03.172311.454eea84a75, nixos-hardware, nixos-unstable-19.09pre175916.1fc591f9a5b"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by andersk**
###### Motivation for this change

This allows Plymouth to show the “NixOS 21.03” label under the logo at startup like it already does at shutdown.

Fixes #59992.

Increases the compressed initrd size by about 5 MB when Plymouth is enabled, since `/etc/plymouth/plugins/label.so` depends on rather more libraries than you’d like. Maybe this is fine since Plymouth is opt-in. We might be able to trim this by using a copy of Cairo compiled against fewer backends? A more minimal version of the label plugin was proposed at https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/45.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by michaelpj** [comment]
Nice! This has been bugging me for literally years.

I wonder if the font should be controlled by some option. What determines the font that gets used during shutdown? Should we try and make sure it's the same one?

**Comment by worldofpeace** [comment]
I'd be fine with using the label-ft patches in plymouth, however... there is a PR that fixes this and overhauls a lot of plymouth https://github.com/NixOS/nixpkgs/pull/84158. But the issue with it is it aspires to do slightly too much in one PR (mostly the commonFunction pieces breaking things)

I'm planning on fixing up that PR and then subsequently defaulting to the bgrt theme. Flicker free boot is blocked on https://github.com/NixOS/nixpkgs/pull/74842 and that is a long ways to being ready.

**Comment by worldofpeace** [comment]
I am however interested in why https://github.com/NixOS/nixpkgs/pull/84158/commits/3128d14a710efbf6cdfc3f3347b65e34291c5aaa is different from what you have...

**Comment by andersk** [comment]
The difference is that 3128d14a710efbf6cdfc3f3347b65e34291c5aaa (#84158) uses the label-ft plugin via a patch (6502fffaddbc9e30a2cb5b38ecb3131d1f04144e) that’s still not merged upstream (https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/45). Plus side: it uses Freetype directly, so it can get away without pulling in 5 MB of libraries and needing a fontconfig configuration. Minus side: there seems to have been no upstream activity toward merging that for 4 years.

**Comment by worldofpeace** [comment]
@jtojnar Does your reservation on using those patches still stand from https://github.com/NixOS/nixpkgs/pull/84158#discussion_r407623151 with this in context? If so, I'd defer to that decision and merge this with a `font` option

**Comment by jtojnar** [comment]
5 MB increase does not sound that bad to me so I would still prefer the upstream label solution over unmerged patches. But initrd maintainers will have final say. We can try to slim down cairo as suggested above it they see it as a problem.

**Review by worldofpeace — CHANGES_REQUESTED** [review]
I actually compared this to my intrd and then a the size increase... not really something to be overly concerned about. It's off by default too.
If we can add a font option like that other plymouth PR I'd merge.

**Comment by andersk** [comment]
Okay, I added an option for the font file.

It’s a little weird in that the *theme* actually decides which font to request from Fontconfig, and this option only affects which font file is made available to Fontconfig during boot to satisfy this request. During shutdown, the normal system Fontconfig configuration and fonts are used.

I set the default as DejaVu Sans because it has the privileged position of being our Fontconfig fallback font:

https://github.com/NixOS/nixpkgs/blob/9816b99e71c3504b0b4c1f8b2e004148460029d4/pkgs/development/libraries/fontconfig/default.nix#L87

If the user adds `noto-fonts` to `config.fonts.fonts`, they’ll actually get Noto Sans during shutdown because that’s what our default `breeze` theme specifies. Those users can configure `boot.plymouth.font = "${pkgs.noto-fonts}/share/fonts/truetype/noto/NotoSans-Regular.ttf"` for a matching font during boot.

File: nixos/modules/system/boot/plymouth.nix
Comment (worldofpeace) [inline nixos/modules/system/boot/plymouth.nix:42]: it would be nice to be able to figure what font is the default sans font for a system. this is just a comment and not a suggestion, but I feel it would be pretty boring to do this with `passthru` listing all the paths to a font within a package, and making a `fonts.fonts.primaryFont`, and fonts having `$variant.path` attr.

**Review by worldofpeace — APPROVED** [review]
I now see the nixos version at boot

File: nixos/modules/system/boot/plymouth.nix
Comment (andersk) [reply nixos/modules/system/boot/plymouth.nix:42]: The relevant font here is not a system default, but one [specified by the theme](https://github.com/KDE/breeze-plymouth/blob/v5.20.5/breeze/breeze.script.cmake#L20-L21).

Finding the default sans font is a simple matter of running `fc-match -f '%{file}' Sans` in an environment with the appropriate Fontconfig configuration. Or if we could somehow pull out the value `Noto Sans` from the selected Plymouth theme, we could run `fc-match -f '%{file}' 'Noto Sans'`. Unfortunately Plymouth themes are written in a custom scripting language and I don’t see a way to run them except in Plymouth itself. We could maybe write some atrocious regex that works most of the time.

**Pull request merged**
